corehq/apps/users/util.py

```diff
@@ -314,60 +314,56 @@ def _last_sync_needs_update(last_sync, sync_datetime):
     return False
 
 
-def log_user_role_update(domain, user_role, user, by_user, updated_via):
-    """
-    :param domain: domain that initiated the change
-    :param user_role: user's new role
-    :param user: couch user that got updated
-    :param by_user: couch user that made the update
-    :param updated_via: web/bulk_importer
-    """
-    message = UserChangeMessage.role_change(user_role)
-    log_user_change(domain, user, by_user, changed_via=updated_via, message=message)
-
-
-def log_user_change(domain, couch_user, changed_by_user, changed_via=None,
-                    message=None, fields_changed=None, action=None,
-                    domain_required_for_log=True, bulk_upload_record_id=None):
+def log_user_change(by_domain, for_domain, couch_user, changed_by_user, changed_via=None,
+                    change_messages=None, fields_changed=None, action=None,
+                    by_domain_required_for_log=True, for_domain_required_for_log=True,
+                    bulk_upload_record_id=None):
     """
     Log changes done to a user.
     For a new user or a deleted user, log only specific fields.
 
-    :param domain: domain where the update was initiated
+    :param by_domain: domain where the update was initiated
+    :param for_domain: domain for which the update was initiated or the domain whose operations will get
+        effected by this change.
+        From user's perspective,
+        A commcare user is completely owned by the domain
+        A web user's membership that let's it perform operations on a domain is owned by the domain.
     :param couch_user: user being changed
     :param changed_by_user: user making the change or SYSTEM_USER_ID
     :param changed_via: changed via medium i.e API/Web
-    :param message: Optional Message text. This message should NEVER be a translated message.
-        Refer https://github.com/dimagi/commcare-hq/pull/30064 for details.
+    :param change_messages: Optional dict of change messages
     :param fields_changed: dict of user fields that have changed with their current value
     :param action: action on the user
-    :param domain_required_for_log: set to False to allow domain less log for specific changes
+    :param by_domain_required_for_log: set to False to allow domain less log for specific changes
+    :param for_domain_required_for_log: set to False to allow domain less log for specific changes
```

Comment:
It is made to be the caller's responsibility to be explicit that it is okay for the log to be without to_ or for_ domain.
The function itself can't really decide for that because it depends on where the logging is happening in the code.

Does that make sense?

Reply:
Yeah, that makes sense, but then why have the check in `log_user_change` at all?  If it's up to the caller to decide whether `for_domain` is appropriate in this context, couldn't the caller just pass in the domain when available and `None` when it isn't?  Are you thinking about scenarios where the caller might have a variable called `domain` but that's actually set to `None` for some reason?

Reply:
The check is to ensure that for_domain is passed.

>  Are you thinking about scenarios where the caller might have a variable called domain but that's actually set to None for some reason?

Yes, something like that where domain could be blank unknowingly. Like places where we are using `request.domain` or other ways to get domain, just a safety net to ensure domain is always passed unless explicitly marked as not required.